internal/jobqueue/fsjobqueue/fsjobqueue.go

```diff
@@ -0,0 +1,340 @@
+// Package fsjobqueue implements a filesystem-backed job queue. It implements
+// the interfaces in package jobqueue.
+//
+// Jobs are stored in the file system, using the `jsondb` package. However,
+// this package does not use the file system as a database, but keeps some
+// state in memory. This means that access to a given directory myst be
+// exclusive to only one `fsJobQueue` object at a time. A single `fsJobQueue`
+// can be safely accessed from multiple goroutines, though.
+//
+// Data is stored non-reduntantly. Any data structure necessary for efficient
+// access (e.g., dependants) are kept in memory.
+package fsjobqueue
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/osbuild/osbuild-composer/internal/jobqueue"
+	"github.com/osbuild/osbuild-composer/internal/jsondb"
+)
+
+type fsJobQueue struct {
+	db *jsondb.JSONDatabase
+
+	// Maps job types to channels of job ids for that type. Only access
+	// through pendingChannel() to ensure concurrent access is restricted
+	// by the mutex.
+	pending      map[string]chan uuid.UUID
+	pendingMutex sync.Mutex
+
+	// Maps job ids to the jobs that depend on it, if any of those
+	// dependants have not yet finished. Only acccess while holding the
+	// associated mutex.
+	dependants      map[uuid.UUID][]uuid.UUID
+	dependantsMutex sync.Mutex
+}
+
+// On-disk job struct. Contains all necessary (but non-redundant) information
+// about a job. These are not held in memory by the job queue, but
+// (de)serialized on each access.
+type job struct {
+	Id           uuid.UUID       `json:"id"`
+	Type         string          `json:"type"`
+	Args         json.RawMessage `json:"args,omitempty"`
+	Dependencies []uuid.UUID     `json:"dependencies"`
+	Result       json.RawMessage `json:"result,omitempty"`
+
+	Status     jobqueue.JobStatus `json:"status"`
+	QueuedAt   time.Time          `json:"queued-at,omitempty"`
```

Comment:
I think we should strive to use underscores in the json format. But really not important, can fix in a follow-up (if at all).

Reply:
Indeed, sorry.